### ISSUE TITLE
suspicious dead code: OPENSSL_VERSION_NUMBER>=0x1010005f (should be >=0x10100004)

### ISSUE THREAD

**Issue by liuqun** [warmcat/libwebsockets#1562]
https://github.com/warmcat/libwebsockets/blob/6d45e6c5b1ac7cf7c0269441ce3a0fae17a07dc0/lib/tls/openssl/ssl.c#L134-L140

https://github.com/openssl/openssl/blob/OpenSSL_1_1_0-pre4/include/openssl/err.h#L368
History preview version of OpenSSL 1.1.0-pre4 (0x10100004) and 1.1.0-pre5 (0x10100005) once changed that prototype defined to `ERR_remove_thread_state(void);`

Then preview version 1.1.0-pre6 (0x10100006) changed it to `ERR_remove_thread_state(void *)`
see https://github.com/openssl/openssl/commit/21e001747d4a246827ad432fc508d67ac404ecdb#diff-9a5faf9ca0ffd5af12ebe8ea072ff981

**Comment by liuqun** [comment]
Suggest:
```
// OpenSSL 1.1.0-pre4/pre5 has a different API prototype (which was later changed back again in pre6)
#if (OPENSSL_VERSION_NUMBER == 0x10100004L || OPENSSL_VERSION_NUMBER == 0x10100005L) && \
    !defined(LIBRESSL_VERSION_NUMBER) && \
    !defined(OPENSSL_IS_BORINGSSL)
    ERR_remove_thread_state();
#endif

// OpenSSL before 1.1.0-pre3 and after 0.9.9(develop branch)
#if (0x00909000L <= OPENSSL_VERSION_NUMBER && \
    OPENSSL_VERSION_NUMBER <= 0x10100003L)
    ERR_remove_thread_state(NULL);
#endif

// openssl before 0.9.9-dev(0x00909000)
// must use old API ERR_remove_state(0)
#if (OPENSSL_VERSION_NUMBER < 0x00909000L) || defined(USE_WOLFSSL) // not tested, but should work...
    ERR_remove_state(0);
#endif
```

**Comment by lws-team** [comment]
Are you sure you followed all the intricacies of OpenSSL version numbering?

https://github.com/openssl/openssl/blob/master/include/openssl/opensslv.h#L131

**Comment by liuqun** [comment]
- https://github.com/openssl/openssl/blob/OpenSSL_1_1_0-pre4/include/openssl/opensslv.h#L33-L38
```
# define OPENSSL_VERSION_NUMBER  0x10100004L
# ifdef OPENSSL_FIPS
#  define OPENSSL_VERSION_TEXT    "OpenSSL 1.1.0-pre4-fips (beta) 16 Mar 2016"
# else
#  define OPENSSL_VERSION_TEXT    "OpenSSL 1.1.0-pre4 (beta) 16 Mar 2016"
# endif
```

- https://github.com/openssl/openssl/blob/OpenSSL_1_1_0/include/openssl/opensslv.h#L42-L47
```
# define OPENSSL_VERSION_NUMBER  0x1010000fL
# ifdef OPENSSL_FIPS
#  define OPENSSL_VERSION_TEXT    "OpenSSL 1.1.0-fips  25 Aug 2016"
# else
#  define OPENSSL_VERSION_TEXT    "OpenSSL 1.1.0  25 Aug 2016"
# endif
```
I am sure that before openssl 1.1.0 release,  in `0xMNN00PPSL`, those guys use `S`=`0xF` as a "Final release". use `S`=`0x1~0xE` as preview pre1~pre15. 

> Are you sure you followed all the intricacies of OpenSSL version numbering?
https://github.com/openssl/openssl/blob/master/include/openssl/opensslv.h#L131

But since openssl 1.1.1 release, the meaning of `S` in `0xMNN00PPSL` has been changed to use `S`=`0x0` as a stable release, rather than the previous `0xF`.
I am confusing about why and when things changed right now.

**Comment by lws-team** [comment]
> I am trying to figure out what should I do in the future since they have changed the definition of 0xMNN00PPSL in the openssl 3.x final release

There's no final release yet for that... I also have been expecting there will be stuff to do with it since every other OpenSSL release in 0.x and 1.x has broken multiple things by their api changes.  I have been trying it at intervals and as of March 20 master built fine against its HEAD of that time with a small api name change fix (45f2c9f9f8ca38e).

Testing OPENSSL_VERSION_NUMBER was how I started out trying to deal with the mess, pretty quickly I switched to empirical api availability tests in cmake to drive selecting the version of the openssl api code use in lws, eg `LWS_HAVE_SSL_CTX_get0_certificate`, rather than try to decode the exact version apis changed in openssl.  If you are building against dev openssl, the public version number it reports is the same for all dev versions (0x.....f), whether that particular dev tree version changed apis yet or not, so trying to test by version cannot properly solve that.

Unfortunately testing for api availability is a bit nuanced too, because apis can switch between being function declarations in the headers or preprocessor defines, which cmake treats separately.  So some things have to be handled by little test programs that cmake tries to build using the desired openssl api and records if they built or not in the `LWS_HAVE...`.

To further increase the difficulty level, you have to make sure that you are not building lws against older (eg, openssl 1.1.1b for Fedora 30) system openssl headers, but only your local 3.x openssl build headers.  Otherwise you will get a mixture of old and new openssl pieces seen by the lws build... check the paths on any C error messages carefully to make sure the headers are not coming out of `/usr/include/openssl`.  I'm using this to build openssl in `/projects/openssl`

```
$ ./config no-tests && make -j32
```

and then this to build lws against the openssl build

```
$ cmake .. -DLWS_WITH_HTTP2=1 -DLWS_OPENSSL_INCLUDE_DIRS=/projects/openssl/include -DLWS_OPENSSL_LIBRARIES="/projects/openssl/libssl.so;/projects/openssl/libcrypto.so" && make -j32
```

You need to delete lws `./build/CMakeCache.txt` if you change openssl version as well, since it caches the results of the openssl api level tests done by lws cmake.

With that I found I needed a two-line patch on CMakeLists.txt to allow it to find a couple of apis on current OpenSSL v3 dev HEAD... please give it a try.
